### PR TITLE
Override headersForRequest (with tests)

### LINKED ISSUE #1003: Ember 2.7 and 2.8 seems to be not sending AUTHORIZATION data
Hi,

When upgrading to ember and ember-data 2.7 and 2.8, it seems like the authorization data is not being sent during requests.

Has anyone else experienced this?

Thanks

### PR DESCRIPTION
**Pull request by f3ndot**
Original work completed by @JeroenvdV in #1009, but now with tests and asserting authorizer presence.

This attempts to fix #1003 so Ember Data 2.7 and greater can send authorization header data when using `DataAdapterMixin`.

@marcoow let me know if this test coverage is sufficient!

### CODE REVIEW

File: addon/mixins/data-adapter-mixin.js [hunk 1]
```diff
@@ -86,6 +89,28 @@ export default Mixin.create({
   },
 
   /**
+    Adds request headers containing the authorization data as constructed
+    by the {{#crossLink "DataAdapterMixin/authorizer:property"}}{{/crossLink}}.
+
+    This method will only called in Ember Data 2.7 or greater, older versions
```
Comment (marcoow) [inline addon/mixins/data-adapter-mixin.js]: `This method will only be called in Ember Data 2.7 or greater, older versions` ;)

File: tests/unit/mixins/data-adapter-mixin-test.js [hunk 2]
```diff
@@ -102,6 +107,88 @@ describe('DataAdapterMixin', () => {
...
+    describe('when the authorizer calls the block', () => {
+      beforeEach(() => {
+        sinon.stub(sessionService, 'authorize', (authorizer, block) => {
+          block('X-Authorization-Header', 'an-auth-value');
+        });
+      });
+
+      it('adds a request header as given by the authorizer', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Authorization-Header');
```
Comment (marcoow) [inline tests/unit/mixins/data-adapter-mixin-test.js]: this is covered by the next line actually

File: tests/unit/mixins/data-adapter-mixin-test.js [hunk 3]
```diff
@@ -102,6 +107,88 @@ describe('DataAdapterMixin', () => {
...
+        expect(headers['X-Authorization-Header']).to.equal('an-auth-value');
+      });
+
+      it('still returns the base headers', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Base-Header');
```
Comment (marcoow) [inline tests/unit/mixins/data-adapter-mixin-test.js]: this is covered by the next line actually

File: tests/unit/mixins/data-adapter-mixin-test.js [hunk 4]
```diff
@@ -102,6 +107,88 @@ describe('DataAdapterMixin', () => {
...
+      it('does not add a request header', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Base-Header');
+        expect(headers).to.not.have.ownProperty('X-Authorization-Header');
+        expect(headers['X-Authorization-Header']).to.not.equal('an-auth-value');
+      });
+
+      it('still returns the base headers', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Base-Header');
```
Comment (marcoow) [inline tests/unit/mixins/data-adapter-mixin-test.js]: this is covered by the next line actually

File: tests/unit/mixins/data-adapter-mixin-test.js [hunk 5]
```diff
@@ -102,6 +107,88 @@ describe('DataAdapterMixin', () => {
...
+    describe('when the authorizer does not call the block', () => {
+      beforeEach(() => {
+        sinon.stub(sessionService, 'authorize');
+      });
+
+      it('does not add a request header', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Base-Header');
+        expect(headers).to.not.have.ownProperty('X-Authorization-Header');
+        expect(headers['X-Authorization-Header']).to.not.equal('an-auth-value');
```
Comment (marcoow) [inline tests/unit/mixins/data-adapter-mixin-test.js]: this is covered by the previous line actually

File: tests/unit/mixins/data-adapter-mixin-test.js [hunk 6]
```diff
@@ -102,6 +107,88 @@ describe('DataAdapterMixin', () => {
...
+    });
+
+    describe('when the authorizer does not call the block', () => {
+      beforeEach(() => {
+        sinon.stub(sessionService, 'authorize');
+      });
+
+      it('does not add a request header', () => {
+        const headers = adapter.headersForRequest();
+        expect(headers).to.have.ownProperty('X-Base-Header');
```
Comment (marcoow) [inline tests/unit/mixins/data-adapter-mixin-test.js]: this is covered by the next test actually

**Comment by marcoow** [comment]
Looks good! Left some comments on minor assertion-style things.

File: addon/mixins/data-adapter-mixin.js (hunk 1 above)
Comment (f3ndot) [inline addon/mixins/data-adapter-mixin.js]: Can you tell I've had 4hrs of sleep today 😉

**Comment by f3ndot** [comment]
@marcoow should be good to go!

**Pull request merged**